src/main/java/org/spongepowered/api/command/args/GenericArguments.java

```diff
@@ -1585,4 +1604,247 @@ public Text getUsage(CommandSource src) {
         }
     }
 
+    /**
+     * Expect an argument to represent a {@link URL}.
+     *
+     * @param key The key to store under
+     * @return the argument
+     */
+    public static CommandElement url(Text key) {
+        return new URIElement(key, false);
+    }
+
+    /**
+     * Expect an argument to represent a {@link URI}.
+     *
+     * @param key The key to store under
+     * @return the argument
+     */
+    public static CommandElement uri(Text key) {
+        return new URIElement(key, true);
+    }
+
+    private static class URIElement extends KeyElement {
+
+        private final boolean returnURI;
+
+        protected URIElement(Text key, boolean returnURI) {
+            super(key);
+            this.returnURI = returnURI;
+        }
+
+        @Nullable
+        @Override
+        protected Object parseValue(CommandSource source, CommandArgs args) throws ArgumentParseException {
+            String str = args.next();
+            URL url;
+            try {
+                url = new URL(str);
+            } catch (MalformedURLException ex) {
+                throw new ArgumentParseException(Text.of("Invalid URL!"), ex, str, 0);
+            }
+            URI uri;
+            try {
+                uri = url.toURI();
+            } catch (URISyntaxException ex) {
+                throw new ArgumentParseException(Text.of("Invalid URL!"), ex, str, 0);
+            }
+            if (this.returnURI) {
```

Comment:
Why do you place the if here? Do you want to verify that it matches the URI format even if it is an URL?
I guess you could move the if before the uri part or wrap the toURI part in the if. Thats just my thought though.